### PR TITLE
[WIP] Lifecycle documentation

### PR DESCRIPTION
**Pull request by wolfbyte**
Documenting Lifecycle interfaces and their usage.

INeedInitialization #779 
IWantToRunBeforeConfigurationIsFinalized #780
IWantToRunWhenConfigurationIsComplete #781
IWantToRunWhenTheBusStartsAndStops #782

### CODE REVIEW

File: nservicebus/lifecycle/iwanttorunwhenthebusstartsandstops.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+Instances are added to the Builder during bus creation using an Instance Per Call lifecyle.
+Instances are all created on the thread responsible for starting the bus. Instances are created with the `Builder` instance and do not require a default constuctor. Instances will have any dependencies injected. 
+Exceptions thrown in the constructors of instances of `IWantToRunWhenTheBusStartsAndStops` are unhandled by NServiceBus. These will bubble up to the code that starts the the bus.
+
+NOTE: The transport is started before any instances of `IWantToRunWhenTheBusStartsAndStops` are created. You cannot rely on all instances of `IWantToRunWhenTheBusStartsAndStops` being created or started before messages start to be processed. 
+
+Once created `Start()` is called on each instance in parallel. Each `Start()` call is made different background thread. Instances of `IWantToRunWhenTheBusStartsAndStops` are kept internally to be stopped when the bus is disposed.
+Exceptions raised from the `Start()` method will cause a [Critical Error](/nservicebus/hosting/critical-errors). As they are run on separate threads, an exception on one `Start()` call will not interfere with any others.
```
Comment (SimonCropp) [inline nservicebus/lifecycle/iwanttorunwhenthebusstartsandstops.md]: should be `[Critical Error](/nservicebus/hosting/critical-errors.md)`

File: nservicebus/lifecycle/iwanttorunwhenthebusstartsandstops.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+title: IWantToRunWhenTheBusStartsAndStops
+summary: An interface that allows you to hook into the startup and shutdown sequence of NServiceBus
+related:
+ - samples/startup-shutdown-sequence
+---
+
+Classes that implement `IWantToRunWhenTheBusStartsAndStops` are invoked just after the bus has been started and just before it is stopped.
+
+Instances are added to the Builder during bus creation using an Instance Per Call lifecyle.
+Instances are all created on the thread responsible for starting the bus. Instances are created with the `Builder` instance and do not require a default constuctor. Instances will have any dependencies injected.
```
Comment (SimonCropp) [inline nservicebus/lifecycle/iwanttorunwhenthebusstartsandstops.md]: "constuctor"

File: nservicebus/lifecycle/iwanttorunbeforeconfigurationisfinalized.md [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+title: IWantToRunBeforeConfigurationIsFinalized
+summary: An interface that allows you to hook into the configuration sequence of NServiceBus
+related:
+ - samples/startup-shutdown-sequence
+---
+
+Classes that implement `IWantToRunBeforeConfigurationIsFinalized` are used to adjust the `Configure` instance which is used to construct the Bus. This occurs just before Feature Activation.
+
+Instances are located during via Type Scanning just before they are invoked. This means that they may be 
+Instances are created as one of the very last steps of Bus creation.
```
Comment (SimonCropp) [inline nservicebus/lifecycle/iwanttorunbeforeconfigurationisfinalized.md]: is this meant to be a new paragraph? if not can u remove the newline

**Comment by SimonCropp** [comment]
r going to add a snippet for each?

File: nservicebus/lifecycle/iwanttorunbeforeconfigurationisfinalized.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+Instances are all created on the calling thread. Instances are created with `Activator.CreateInstance(...)` and require a default constructor. Instances will not have any dependencies injected.
```
Comment (SimonCropp) [inline nservicebus/lifecycle/iwanttorunbeforeconfigurationisfinalized.md]: make `Activator.CreateInstance` link to msdn

File: nservicebus/lifecycle/ineedinitialization.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+summary: An interface that allows you to hook into the very beginning of the bus creation sequence of NServiceBus
+tags:
+ - lifecycle
+related:
+ - samples/startup-shutdown-sequence
+---
+
+Classes that implement `NServiceBus.INeedInitialization` are created and called as one of the first steps in the bus creation lifecycle. Use `INeedInitialization` to register components that will be used later in the bus creation lifecycle. 
+
+NOTE: In V3 of NServiceBus, this interface is found in the `NServiceBus.Config` namespace. In V4 both interfaces are available but the old one is marked as obsolete. As of V5 use of the `NServiceBus.Config.INeedInitialization` interface will cause a compile-time error. In V4 (which has both interfaces), all instances of `NServiceBus.Config.INeedInitialization` are created and called and then all instances of `NServiceBus.INeedInitialization` are created and called.
```
Comment (SimonCropp) [inline nservicebus/lifecycle/ineedinitialization.md]: i know this is minor but everywhere else in the doco we use "Version X" not "VX"

File: nservicebus/lifecycle/iwanttorunwhenconfigurationiscomplete.md [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+---
+
+Classes that implement `IWantToRunWhenConfigurationIsComplete` are executed right before the bus is created.
+
+Instances are located during Type Scanning. Alternatively they may be added by instances of `INeedInitialization` or `IWantToRunBeforeConfigurationIsFinalized`.
+Instances are all created on the thread responsible for creating the bus. Instances are created with the `Builder` instance and do not require a default constructor. Instances will have any dependencies injected. 
+Once created `Run(Configure)` is called on each one in series. 
+Exceptions thrown by instances of `IWantToRunWhenConfigurationIsComplete` are unhandled by NServiceBus. These will bubble up to the creator of the bus.
+
+Use `IWantToRunWhenConfigurationIsComplete` for any tasks that need to be run at the end of bus creation lifecycle. This might include diagnostic information about configuration or initialization that downstream Features and components might depend on.
```
Comment (timbussmann) [inline nservicebus/lifecycle/iwanttorunwhenconfigurationiscomplete.md]: Should we add the warning that this interface is marked as obsolete with v6 and refer to the Feature docu? (which I'll add soon...)

**Comment by wolfbyte** [comment]
Snippets added. For those features that were not present in 3 or have been deprecated in 6 should there be snippets for those versions indicating that they are not available?

**Pull request merged**